// actionhero/dist/modules/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = void 0;
const arrayStartingMatch_1 = require("./utils/arrayStartingMatch");
const arrayUnique_1 = require("./utils/arrayUnique");
const asyncWaterfall_1 = require("./utils/asyncWaterfall");
const collapseObjectToArray_1 = require("./utils/collapseObjectToArray");
const ensureNoTsHeaderFiles_1 = require("./utils/ensureNoTsHeaderFiles");
const eventLoopDelay_1 = require("./utils/eventLoopDelay");
const filterObjectForLogging_1 = require("./utils/filterObjectForLogging");
const getExternalIPAddress_1 = require("./utils/getExternalIPAddress");
const hashMerge_1 = require("./utils/hashMerge");
const isPlainObject_1 = require("./utils/isPlainObject");
const parseHeadersForClientAddress_1 = require("./utils/parseHeadersForClientAddress");
const parseCookies_1 = require("./utils/parseCookies");
const parseIPv6URI_1 = require("./utils/parseIPv6URI");
const replaceDistWithSrc_1 = require("./utils/replaceDistWithSrc");
const sleep_1 = require("./utils/sleep");
const sortGlobalMiddleware_1 = require("./utils/sortGlobalMiddleware");
const sourceRelativeLinkPath_1 = require("./utils/sourceRelativeLinkPath");
const fileUtils_1 = require("./utils/fileUtils");
/**
 * Utility functions for Actionhero
 */
exports.utils = {
    arrayStartingMatch: arrayStartingMatch_1.arrayStartingMatch,
    arrayUnique: arrayUnique_1.arrayUnique,
    asyncWaterfall: asyncWaterfall_1.asyncWaterfall,
    collapseObjectToArray: collapseObjectToArray_1.collapseObjectToArray,
    ensureNoTsHeaderFiles: ensureNoTsHeaderFiles_1.ensureNoTsHeaderFiles,
    eventLoopDelay: eventLoopDelay_1.eventLoopDelay,
    filterObjectForLogging: filterObjectForLogging_1.filterObjectForLogging,
    getExternalIPAddress: getExternalIPAddress_1.getExternalIPAddress,
    hashMerge: hashMerge_1.hashMerge,
    isPlainObject: isPlainObject_1.isPlainObject,
    parseHeadersForClientAddress: parseHeadersForClientAddress_1.parseHeadersForClientAddress,
    parseCookies: parseCookies_1.parseCookies,
    parseIPv6URI: parseIPv6URI_1.parseIPv6URI,
    replaceDistWithSrc: replaceDistWithSrc_1.replaceDistWithSrc,
    sleep: sleep_1.sleep,
    sortGlobalMiddleware: sortGlobalMiddleware_1.sortGlobalMiddleware,
    sourceRelativeLinkPath: sourceRelativeLinkPath_1.sourceRelativeLinkPath,
    fileUtils: {
        dirExists: fileUtils_1.dirExists,
        fileExists: fileUtils_1.fileExists,
        createDirSafely: fileUtils_1.createDirSafely,
        createFileSafely: fileUtils_1.createFileSafely,
        createLinkfileSafely: fileUtils_1.createLinkfileSafely,
        removeLinkfileSafely: fileUtils_1.removeLinkfileSafely,
        createSymlinkSafely: fileUtils_1.createSymlinkSafely,
    },
};
