// actionhero/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.api = void 0;
// export classes (capitalized)
var api_1 = require("./classes/api");
Object.defineProperty(exports, "Api", { enumerable: true, get: function () { return api_1.Api; } });
var process_1 = require("./classes/process");
Object.defineProperty(exports, "Process", { enumerable: true, get: function () { return process_1.Process; } });
var initializer_1 = require("./classes/initializer");
Object.defineProperty(exports, "Initializer", { enumerable: true, get: function () { return initializer_1.Initializer; } });
var connection_1 = require("./classes/connection");
Object.defineProperty(exports, "Connection", { enumerable: true, get: function () { return connection_1.Connection; } });
var action_1 = require("./classes/action");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return action_1.Action; } });
var task_1 = require("./classes/task");
Object.defineProperty(exports, "Task", { enumerable: true, get: function () { return task_1.Task; } });
var server_1 = require("./classes/server");
Object.defineProperty(exports, "Server", { enumerable: true, get: function () { return server_1.Server; } });
var cli_1 = require("./classes/cli");
Object.defineProperty(exports, "CLI", { enumerable: true, get: function () { return cli_1.CLI; } });
var actionProcessor_1 = require("./classes/actionProcessor");
Object.defineProperty(exports, "ActionProcessor", { enumerable: true, get: function () { return actionProcessor_1.ActionProcessor; } });
// export modules (lower case)
var utils_1 = require("./modules/utils");
Object.defineProperty(exports, "utils", { enumerable: true, get: function () { return utils_1.utils; } });
var config_1 = require("./modules/config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_1.config; } });
var log_1 = require("./modules/log");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return log_1.log; } });
var action_2 = require("./modules/action");
Object.defineProperty(exports, "action", { enumerable: true, get: function () { return action_2.action; } });
var task_2 = require("./modules/task");
Object.defineProperty(exports, "task", { enumerable: true, get: function () { return task_2.task; } });
var cache_1 = require("./modules/cache");
Object.defineProperty(exports, "cache", { enumerable: true, get: function () { return cache_1.cache; } });
var chatRoom_1 = require("./modules/chatRoom");
Object.defineProperty(exports, "chatRoom", { enumerable: true, get: function () { return chatRoom_1.chatRoom; } });
var redis_1 = require("./modules/redis");
Object.defineProperty(exports, "redis", { enumerable: true, get: function () { return redis_1.redis; } });
var i18n_1 = require("./modules/i18n");
Object.defineProperty(exports, "i18n", { enumerable: true, get: function () { return i18n_1.i18n; } });
var route_1 = require("./modules/route");
Object.defineProperty(exports, "route", { enumerable: true, get: function () { return route_1.route; } });
var specHelper_1 = require("./modules/specHelper");
Object.defineProperty(exports, "specHelper", { enumerable: true, get: function () { return specHelper_1.specHelper; } });
// export static members of this process (lower case)
var env_1 = require("./classes/process/env");
Object.defineProperty(exports, "env", { enumerable: true, get: function () { return env_1.env; } });
var actionheroVersion_1 = require("./classes/process/actionheroVersion");
Object.defineProperty(exports, "actionheroVersion", { enumerable: true, get: function () { return actionheroVersion_1.actionheroVersion; } });
var projectRoot_1 = require("./classes/process/projectRoot");
Object.defineProperty(exports, "projectRoot", { enumerable: true, get: function () { return projectRoot_1.projectRoot; } });
var typescript_1 = require("./classes/process/typescript");
Object.defineProperty(exports, "typescript", { enumerable: true, get: function () { return typescript_1.typescript; } });
var id_1 = require("./classes/process/id");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return id_1.id; } });
// API object to hold connections, actions, tasks, initializers, and servers
const api_2 = require("./classes/api");
// backwards-compatibility for older versions of node.js
// we can't use globalThis for node v8, v10
// @ts-ignore
if (!global.api) {
    // @ts-ignore
    global.api = new api_2.Api();
}
// @ts-ignore
exports.api = global.api;
